Add `rawtxwithfee` ZMQ topic handling

0xB10C opened the pull request:
This PR adds the handling of the `rawtxwithfee` ZMQ topic. Messages with this topic are published by a customized Bitcoin Core instance. Next to a lot of refactors and cleanups, the integration test environment setup is moved to the Makefile. A `bitcoind` instance running in a Docker container is used now. Since the GitHub Actions CI seems to be quite unstable and error prone I moved to Travis CI.

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
+}
+
+// Stop sets the cancel flag to true. The ZMQSubscriber is stopped after it finishes
+// receiving a message or
```
`or reaches the timeout`

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
+
+func parseTransaction(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
```
unbalanced parens

OttoAllmendinger replied:
still unbalanced ;)

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
+	fee := binary.BigEndian.Uint64(feeBytes)
+
+	vsizeFloat := (float64(wireTx.SerializeSizeStripped()*3) + float64(wireTx.SerializeSize())) / 4
+	vsize := int(math.Ceil(vsizeFloat))
```
how about we track the weight units instead and keep full precision?

0xB10C replied:
Good point!

0xB10C replied:
Changed in https://github.com/0xB10C/bademeister-go/pull/19/commits/08d2148b10a7831418f37b30d78a5e1ae872d65d

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
+	if length <= 8 {
+		return nil, errors.New("unexpected rawtxwithfee length")
+	}
+	rawtx, feeBytes := rawtxwithfee[:length-8], rawtxwithfee[length-8:]
```
I'm not familiar with how it is serialized in bitcoin core, but can we not add this as a separate `payload` element?

0xB10C replied:
Do you mean a struct?

```go
type payload struct {
	rawtx    []byte,
	feeBytes [8]byte,
}
```

OttoAllmendinger replied:
I meant putting it between `payload[0]` and `payload[1]` but we clarified out-of-band

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -52,32 +52,54 @@ func getTestZMQSubscriber(t *testing.T, env test.TestEnv) *ZMQSubscriber {
+	}
+
+	// Generate 101 blocks to have spendable UTXOs.
+	env.GenerateToAddress(101, addressSendFrom)
```
I'm ok with removing `panic()` for recoverable errors, but we have to do the err checks then

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -52,32 +52,54 @@ func getTestZMQSubscriber(t *testing.T, env test.TestEnv) *ZMQSubscriber {
+	env.GenerateToAddress(1, addressSendFrom)
 	require.NotNil(t, waitForZMQBlock(t, z, waitTimeout))
 
+	env.SendSimpleTransaction(addressSendTo)
```
err check (I'm ok with `require.NoError`)

0xB10C commented:
Thanks for the review. Will address this after I'm happy with the integration tests.

0xB10C commented on src/test/testenv.go:
```diff
@@ -78,61 +81,128 @@ func newBitcoindRpc() (*rpcclient.Client, error) {
+}
+
+// GetNewAddress returns a new address via the `getnewaddress` RPC.
+func (e *TestEnv) GetNewAddress() (btcutil.Address, error) {
```
Somehow didn't see that... Thanks!

0xB10C commented:
I think this is finally ready for review after a long battle with Github Action CI and Travis CI. I've updated the initial description to summarize what I've changed.

OttoAllmendinger commented on src/bitcoinrpcclient/bitcoinrpcclient.go:
```diff
@@ -0,0 +1,167 @@
+// WaitTillRPCServerReady tries every `checkInterval` if the RPC Server the
+// client is connected to is ready. It returns nil if the server is ready, or
+// otherwise times out after `timeout`.
+func (rpcClient *BitcoinRPCClient) WaitTillRPCServerReady(timeout time.Duration) error {
```
why not call this in the constructor instead? less work for the user

OttoAllmendinger commented on src/storage/storage.go:
```diff
@@ -37,37 +39,54 @@ func NewStorage(path string) (*Storage, error) {
+		return fmt.Errorf("could not fill the `config` table: %s", err)
+	}
+
+	const crateMempoolTxTable string = `
```
`create`

OttoAllmendinger commented on src/storage/storage.go:
```diff
@@ -130,16 +151,22 @@ func (i *TxIterator) Next() *types.Transaction {
+	var fee uint64
+	var weight int
+	if err := i.rows.Scan(&txidBytes, &firstSeen, &fee, &weight); err != nil {
+		panic(err) // TODO: cleanup
```
IMO `panic()` is fine here since the caller can only propagate anyway

0xB10C replied:
Agree, will remove it.

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +118,98 @@ func (z *ZMQSubscriber) Run() error {
+		if err != nil {
+			return err
+		}
+		z.IncomingTx <- *tx
```
we can prob get rid of that later

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
+
+		if currentTimeout > maxTimeout {
+			fmt.Printf("Timeout (%s) while waiting for '%s'.\n", maxTimeout, cookiePath)
+			os.Exit(1)
```
`error` return would be a bit more standard

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
 
-	z := getTestZMQSubscriber(t, env)
+	z, err := setupAndRunZMQSubscriber(t, zmqHost, zmqPort)
 	defer z.Stop()
```
should come after `require.NoError`

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
 	{
-		z2 := getTestZMQSubscriber(t, env)
+		z2, err := setupAndRunZMQSubscriber(t, zmqHost, zmqPort)
 		defer z2.Stop()
```
after `require.NoError()`

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -59,92 +46,68 @@ func NewZMQSubscriber(host string, port string) (*ZMQSubscriber, error) {
 		}
 
 		topic, payload := string(msg[0]), msg[1:]
-		log.Printf(`[ZMQ] received topic "%s"`, topic)
```
the idea here was to have prefixes for different subsystems that can be easily grepped

0xB10C replied:
Could use https://golang.org/pkg/log/#SetPrefix or set the `Lshortfile` flag (https://golang.org/pkg/log/#pkg-constants) for a short filename in the log.

OttoAllmendinger commented on src/storage/storage.go:
```diff
@@ -3,15 +3,17 @@ package storage
 )
 
-const VERSION = 2
+const currentVersion = 4
```
IMO it is fine to bump more than 1 version per PR but not required

0xB10C replied:
Agree for future PRs, won't change it back now..

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
+		t.Skip("Skipping " + t.Name() + " since it's not a unit test.")
+	}
+
+	const zmqWaitTimeout = 5 * time.Second
```
nit: if always we pass the same timeout anyway (and don't want to make it a global const either) we can make it a variable that is local to the func

0xB10C replied:
I think that's what it is now, isn't it? It's just `const` because it should not be modified.

OttoAllmendinger replied:
I meant local to `waitForForZMQBlock`

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
 func TestZMQSubscriber(t *testing.T) {
-	const waitTimeout = 5 * time.Second
+	if testing.Short() {
+		t.Skip("Skipping " + t.Name() + " since it's not a unit test.")
```
I'm a bit surprised that we don't have to return here, but it looks correct https://golang.org/pkg/testing/#hdr-Skipping

does it call `os.Exit()` then?

0xB10C replied:
Tests are run in goroutines. `Skip()` exits the current goroutine and marks the tests as "skipped".

OttoAllmendinger replied:
but how does `t.Skip()` exit the parent goroutine?

0xB10C replied:
with `runtime.Goexit()`. See https://golang.org/pkg/testing/#B.SkipNow

OttoAllmendinger replied:
TIL!

OttoAllmendinger commented on src/storage/storage_test.go:
```diff
@@ -44,16 +43,21 @@ func testQueryTransactions(t *testing.T, st *Storage, firstSeen time.Time, txs [
 	}
 
-	// create from empty file
+	// The environment variable `TEST_INTEGRATION_DIR` is set to a temporay directory created by
```
temporary

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +118,98 @@ func (z *ZMQSubscriber) Run() error {
+	// which is not used here, but noted for completeness.
+	// rawtxwithfee is the rawtx by Bitcoin Core concatinated with the 8 byte BE
+	// transaction fee
+	rawtxwithfee, _ := payload[0], payload[1]
```
pls add a link to the source

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber_test.go:
```diff
@@ -36,48 +54,126 @@ func waitForZMQBlock(t *testing.T, z *ZMQSubscriber, timeout time.Duration) *typ
+	const maxTimeout = 10 * time.Second
+	const timeoutDelay = 250 * time.Millisecond
+
+	// The environment variable `TEST_INTEGRATION_DIR` is set to a temporay directory created by
```
`temporary`

OttoAllmendinger commented:
Please consider #20 instead of providing the custom bitcoind binary for tests. We shouldn't ask potential contributors to run opaque binaries.

0xB10C commented:
Agree. Will have a look at building a docker image with the patched bitcoind next. Is that in the scope of this PR?

OttoAllmendinger commented:
I think it should be a different PR, but if we can get it in before this one we can avoid having the binary in the git history

0xB10C commented:
Addressed some of the outstanding comments.

OttoAllmendinger commented on src/zmqsubscriber/zmqsubscriber.go:
```diff
@@ -155,6 +118,99 @@ func (z *ZMQSubscriber) Run() error {
+
+func parseTransactionHash(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
```
`expected len([txhash, sequence]) == 2`

0xB10C replied:
Maybe I keep fixing up the wrong commit? Changed this three times now...

0xB10C replied:
Oh I see. It's in `parseTransactionHash`. Going to remove that one in an extra commit.

0xB10C commented:
Removed the `bitcoind` binary. Now uses a patched `bitcoind` running inside a Docker container based on this image https://hub.docker.com/r/b10c/bitcoind-patched-zmq.
